### PR TITLE
Python 3.10 / Django 3.2 upgrade

### PR DESCRIPTION
**Pull request by gumaerc**
# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1854

# Description (What does it do?)
This PR primarily upgrades the version of Python used in the project from 3.8.7 to 3.10.12. Some other libraries have been upgraded to be compatible with this new Python version:

 - Django and all the 3rd party Django apps
 - `concoursepy` was swapped out with our forked version
 - `black`, `pylint` and `pytest`
 - `ol-django` libraries were updated to the new versions released after https://github.com/mitodl/ol-django/pull/125 to support using a new version of `pytest` in dependent projects

In addition to the updated libraries, there was some other work that needed to be done:
 - The Github actions configuration was also updated with the new Python version and some changes were made to support the new versions of `pylint` and `pytest` being used
 - Docstrings that have a leading and trailing space in the triple quotes were automatically reformatted by the new version of `black`. This lead to a lot of noise in the change list for this PR, a lot of the changed files are simply removing extra whitespace from docstrings.
 - Some syntax was updated based on recommendations in the new version of `pylint`, such as:
   - Using dict literals (`{"key": value}`) instead of `dict(key=value)`
   - Using `f` strings instead of `.format`
   - Specifying `encoding="utf-8"` when opening files
   - using `str(value)` instead of `value.__str__()`
   - Removing unnecessary parentheses
 - Some other syntax changes from new versions of Django apps, such as:
   - `{% load hijack %}` instead of `{% load hijack_tags %}`, other changes related to `django-hijack`
   - `from django.conf.urls import url` -> `from django.urls import re_path`
 - Some tests had to be reworked to support the new version of pytest

# How can this be tested?
Before getting started, you should rebuild your containers with `docker-compose build --no-cache` to be certain that you are working with containers built with the new versions of everything. Beyond that, the entire site should be smoke tested. This is a basic list I came up with, but anything else should be added here as necessary:

 - [ ] Django admin
 - [ ] pylint / pytest / black
 - [ ] Editing content and saving it
   - [ ] Pages
   - [ ] Resources
   - [ ] Metadata
   - [ ] Menus
 - [ ] Pushing up pipelines to Concourse
   - [ ] Creating a new site
   - [ ] `backpopulate_pipelines`
   - [ ] `upsert_theme_assets_pipline`
   - [ ] `upsert_mass_build_pipeline`
 - [ ] Google drive sync
 - [ ] Youtube
 - [ ] 3Play
 - [ ] AWS S3
 - [ ] AWS MediaConvert

### CODE REVIEW

**Comment by gumaerc** [comment]
Remaining to fix:

 - `pylint`
 - `pytest`

**Comment by gumaerc** [comment]
@HussainTaj-arbisoft This PR is all rebased and ready for you to take over. Once https://github.com/mitodl/ol-django/pull/125 is merged, we should be able to use updated `pytest`. The work beyond then is making sure tests pass and manually testing all the things I mentioned in https://github.com/mitodl/ocw-studio/issues/1854.

**Comment by rhysyngsun** [comment]
@HussainTaj-arbisoft the fixed version of `mitol-django-common` is available now: https://pypi.org/project/mitol-django-common/2023.6.27.1/

**Comment by gumaerc** [comment]
Some caveats:

 - After upgrading the `ol-django` libraries after @rhysyngsun merged https://github.com/mitodl/ol-django/pull/125 and released the new versions, I was able to install the new version of `pytest` but one of the tests (the only one being actually imported from a helper in `ol-django`) is not working properly. The test is `mitol.common.pytest_utils.test_app_json_modified`. From what I understand, it's basically supposed to compare your current `app.json` with what would be generated by the `generate_app_json` management command, and if there are differences the tests fail which forces you to keep `app.json` up to date in order to merge anything. When I run the management command manually or step into a shell and run the functions there, the config evaluates as the same. For some reason inside the `pytest` environment, the exception block indicating that they are not the same is being hit and I wasn't able to figure out why. I did a bunch of debugging yesterday, but as the code being run is inside of an imported library installed in the docker container I had a hard time digging in and getting clues on exactly why they might evaluate differently in the test, but not when running the management command.
 - I had to enable `ignore: Exception ignored in:pytest.PytestUnraisableExceptionWarning` in `pytest.ini`, as I was getting flaky test failures with 2 tests that had to do with transcript handling that were apparently leaving files hanging open. I dug into the tests and they're not using actual files, they're using `BytesIO` objects in memory as far as I can tell. I tried calling `.close()` on the transcript files before the end of the tests and that had no effect. The only way I could get the tests to pass was by adding the ignore line to the config.

**Comment by mbertrand** [comment]
Regarding this issue:

> I was getting flaky test failures with 2 tests that had to do with transcript handling that were apparently leaving files hanging open

I ran into that with open-discussions, and needed to modify blocks of code where files were being opened/read by putting them inside a `with` block:  https://github.com/mitodl/open-discussions/pull/4019/files#diff-d2ee8a31b0462e1a17144737d6f0eb91b30cfb6916c0949d8de74582e096f9f4L144-R150

File: .github/workflows/ci.yml
Comment (mbertrand) [inline .github/workflows/ci.yml:5]: Might be best to keep this at the same version as the heroku stack (currently 20) to avoid any surprises - https://github.com/mitodl/mitxpro/issues/2496

```
-----> Building on the Heroku-20 stack
```

File: main/settings.py
Comment (mbertrand) [inline main/settings.py:135]: Not seeing the "hijack" option for users in django admin.  Not sure why, all the pieces seem to be in place.

File: users/admin.py
Comment (mbertrand) [inline users/admin.py:5]: The hijack button doesn't show up for me unless I remove this import and it's usage below.

**Review by mbertrand — CHANGES_REQUESTED** [review]
A few preliminary things, waiting on a fresh import of the database to finish over lunchtime then I'll do more testing.

**Review by mbertrand — COMMENTED** [review]
Getting an error when syncing to github on this branch (still works okay on master):

```
celery_1          | [2023-07-07 17:56:14,344: ERROR/ForkPoolWorker-8] Task content_sync.tasks.sync_website_content[bd9abc86-a8e0-4acb-b9cc-4724fe2a9b52] raised unexpected: AssertionError('supplied private key should be a string')
celery_1          | Traceback (most recent call last):
celery_1          |   File "/usr/local/lib/python3.10/site-packages/celery/app/trace.py", line 477, in trace_task
celery_1          |     R = retval = fun(*args, **kwargs)
celery_1          |   File "/usr/local/lib/python3.10/site-packages/sentry_sdk/integrations/celery.py", line 275, in _inner
celery_1          |     reraise(*exc_info)
celery_1          |   File "/usr/local/lib/python3.10/site-packages/sentry_sdk/_compat.py", line 60, in reraise
celery_1          |     raise value
celery_1          |   File "/usr/local/lib/python3.10/site-packages/sentry_sdk/integrations/celery.py", line 270, in _inner
celery_1          |     return f(*args, **kwargs)
celery_1          |   File "/usr/local/lib/python3.10/site-packages/celery/app/trace.py", line 760, in __protected_call__
celery_1          |     return self.run(*args, **kwargs)
celery_1          |   File "/usr/local/lib/python3.10/site-packages/celery/app/autoretry.py", line 38, in run
celery_1          |     return task._orig_run(*args, **kwargs)
celery_1          |   File "/src/content_sync/decorators.py", line 100, in wrapper
celery_1          |     return_value = func(*args, **kwargs)
celery_1          |   File "/src/content_sync/tasks.py", line 220, in sync_website_content
celery_1          |     backend = api.get_sync_backend(website)
celery_1          |   File "/src/content_sync/api.py", line 35, in get_sync_backend
celery_1          |     return import_string(settings.CONTENT_SYNC_BACKEND)(website)
celery_1          |   File "/src/content_sync/backends/github.py", line 39, in __init__
celery_1          |     self.api = GithubApiWrapper(self.website, self.site_config)
celery_1          |   File "/src/content_sync/apis/github.py", line 183, in __init__
celery_1          |     login_or_token=get_token(),
celery_1          |   File "/src/content_sync/apis/github.py", line 148, in get_token
celery_1          |     app = GithubIntegration(
celery_1          |   File "/usr/local/lib/python3.10/site-packages/github/GithubIntegration.py", line 38, in __init__
celery_1          |     assert isinstance(private_key, str), "supplied private key should be a string"
celery_1          | AssertionError: supplied private key should be a string
```

**Comment by gumaerc** [comment]
@mbertrand Are you sure your Celery container was also rebuilt? It looks like it's using the old version of the `pygithub` library.

**Comment by mbertrand** [comment]
> @mbertrand Are you sure your Celery container was also rebuilt? It looks like it's using the old version of the `pygithub` library.

I ran `docker-compose build --no-cache web celery`.  Will do it again.

**Comment by mbertrand** [comment]
Deleted existing celery container, ran `docker-compose build --no-cache celery`, still getting the same error.  The pygithub version seems correct:

```
docker-compose run --rm celery bash
Creating ocw-studio_celery_run ... done
mitodl@40b657f0bcbd:/src$ pip show pygithub
Name: PyGithub
Version: 1.58.2
Summary: Use the full Github API v3
Home-page: https://github.com/pygithub/pygithub
Author: Vincent Jacques
Author-email: vincent@vincent-jacques.net
License:
Location: /usr/local/lib/python3.10/site-packages
Requires: deprecated, pyjwt, pynacl, requests
Required-by:
mitodl@40b657f0bcbd:/src$ exit
exit
```

I have the following set in my .env file:
```
GIT_ORGANIZATION=ocwdev-mb
GITHUB_APP_ID=<app_id>
GITHUB_APP_PRIVATE_KEY==-----BEGIN RSA PRIVATE KEY-----\nMIIE...........\n-----END RSA PRIVATE KEY-----
```

If I comment out `GITHUB_APP_ID` and `GITHUB_APP_PRIVATE_KEY`, and set a valid value for `GIT_API_TOKEN` instead, then it works.  But RC and prod use the former pair of variables.

**Comment by gumaerc** [comment]
@mbertrand I see, so this is only an issue on MIT's enterprise Github where private key authentication is required. Github integration using `GIT_TOKEN` and github.com seems to work fine, so maybe something changed. I'll look into it.

**Comment by gumaerc** [comment]
@mbertrand This should be ready for another look

**Review by mbertrand — APPROVED** [review]
Everything seems to work well now, only thing I can think of that I didn't test yet is 3Play integration.  I might be able to get around to that later tomorrow but I'll approve this now if you'd rather get it out the door and have 3Play tested on RC.

**Comment by gumaerc** [comment]
@mbertrand Let's test that tomorrow in RC, it will be easier there with YT / 3Play all set up and nothing will need to be spoofed.

**Pull request merged**
